### PR TITLE
Learn class balance using tensor product algorithm

### PR DESCRIPTION
**Pull request by ajratner**
@bhancock8 Only a quick look if you have time!

@fredsala might be interested to check this out!

### CODE REVIEW

**Comment by ajratner** [comment]
Three high-level notes:
- `einsum` is awesome
- I had trouble getting it to work with 1st order approaches i.e. gradient descent... but it's not that big of a model, so I tried L-BFGS- it works fine, so I left that in for now.  Obviously there are other approaches for tensor product factorization that would be faster, but L-BFGS is already in `torch`, really simple, and seems to work just fine! 
- At the end, I break col-wise symmetry by getting the recovered LF accuracy estimate, and assuming that on avg these are better than random

File: metal/label_model/class_balance.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    Note: This approach can also be used for estimation of the LabelModel, may
+    want to later refactor and expand this class.
+    """
+
+    def __init__(self, k, abstains=True, config=None):
+        super().__init__()
+        self.config = config
+        self.k = k
+        self.abstains = abstains
+        self.k_0 = 0 if self.abstains else 1
```
Comment (bhancock8) [inline metal/label_model/class_balance.py]: A couple of definitions here for self.k_0 and self.k_lf would be nice, since k can mean a lot of very-similar-but-different-in-important-ways things. Either in the class docstring or here inline.

File: metal/label_model/class_balance.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
+from itertools import product
+
+import numpy as np
+import torch
+from torch import nn, optim
+
+
+class ClassBalanceModel(nn.Module):
```
Comment (bhancock8) [inline metal/label_model/class_balance.py]: Especially since this is under the label_model/ directory, I think it's probably worth an extra disclaimer in the docstring that this is _not_ a LabelModel. You do very clearly define what it's actually computing, but I'm sure someone will try to use it the wrong way otherwise.

File: metal/label_model/class_balance.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    @staticmethod
+    def get_loss(O, Q, mask):
+        # Main constraint: match empirical three-way overlaps matrix
+        # (entries O_{ijk} for i != j != k)
+        diffs = (O - torch.einsum("aby,cdy,efy->acebdf", [Q, Q, Q]))[mask]
+        return torch.norm(diffs) ** 2
+
+    def train_model(self, L=None, O=None, lr=1, max_iter=1000, verbose=False):
+
```
Comment (bhancock8) [inline metal/label_model/class_balance.py]: Drop the extra blank line

File: metal/label_model/class_balance.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    def _get_overlaps_tensor(self, L):
+        """Transforms the input label matrix to a three-way overlaps tensor.
+
+        Args:
+            L: (np.array) An n x m array of LF output labels, in {0,...,k} if
+                self.abstains, else in {1,...,k}, generated by m conditionally
+                independent LFs on n data points
+
+        Outputs:
+            O: (torch.Tensor) A (m, m, m, k, k, k) tensor of the label-specific
```
Comment (bhancock8) [inline metal/label_model/class_balance.py]: Wow--a 6D tensor. May be worth saying something like, "for example, O[a,b,c,d,e,f] corresponds to...[explanation that uses a-f]"

File: metal/label_model/class_balance.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            O = self._get_overlaps_tensor(L)
+        else:
+            raise ValueError("L or O required as input.")
+        self.m = O.shape[0]
+
+        # Compute mask
+        self.mask = self.get_mask(self.m)
+
+        # Initialize parameters
+        self.Q = nn.Parameter(
```
Comment (bhancock8) [inline metal/label_model/class_balance.py]: That is a whole lot of converting and wrapping. Does torch.random not work here?

File: metal/label_model/class_balance.py (hunk 1 above)
Comment (ajratner) [reply metal/label_model/class_balance.py]: Done!

File: metal/label_model/class_balance.py (hunk 2 above)
Comment (ajratner) [reply metal/label_model/class_balance.py]: It actually is one, as it implicitly estimates the accuracies of the LFs, *for conditionally independent LFs.*  How exactly it fits into the overall repo, therefore, may be a little bit fluid... but I want to get it in at least in this separate form now, so can at least be used for class balance estimation

File: metal/label_model/class_balance.py (hunk 3 above)
Comment (ajratner) [reply metal/label_model/class_balance.py]: Done

File: metal/label_model/class_balance.py (hunk 4 above)
Comment (ajratner) [reply metal/label_model/class_balance.py]: Done!

File: metal/label_model/class_balance.py (hunk 5 above)
Comment (ajratner) [reply metal/label_model/class_balance.py]: Fixed!
